File: src/core/detectors/MockDetectorAdapter.js
import { DetectorAdapter } from './DetectorAdapter'

const SCENARIOS = {
  one_person: {
    label: '1 người',
    objects: [
      {
        trackId: 'T-001',
        class: 'person',
        confidence: 0.94,
        bbox: { x: 0.35, y: 0.2, width: 0.12, height: 0.45 },
        attributes: { helmet: true, mask: true, coverall: true },
      },
    ],
  },
  three_persons: {
    label: '3 người',
    objects: [
      {
        trackId: 'T-101',
        class: 'person',
        confidence: 0.91,
        bbox: { x: 0.1, y: 0.25, width: 0.1, height: 0.4 },
        attributes: { helmet: true, mask: false, coverall: true },
      },
      {
        trackId: 'T-102',
        class: 'person',
        confidence: 0.88,
        bbox: { x: 0.35, y: 0.22, width: 0.11, height: 0.42 },
        attributes: { helmet: true, mask: true, coverall: true },
      },
      {
        trackId: 'T-103',
        class: 'person',
        confidence: 0.86,
        bbox: { x: 0.62, y: 0.28, width: 0.1, height: 0.38 },
        attributes: { helmet: false, mask: true, coverall: true },
      },
    ],
  },
  one_animal: {
    label: '1 động vật',
    objects: [
      {
        trackId: 'T-AN-01',
        class: 'animal',
        confidence: 0.89,
        bbox: { x: 0.45, y: 0.55, width: 0.18, height: 0.25 },
        attributes: {},
      },
    ],
  },
  ppe_violation: {
    label: 'Vi phạm PPE',
    objects: [
      {
        trackId: 'T-PPE-01',
        class: 'person',
        confidence: 0.93,
        bbox: { x: 0.4, y: 0.18, width: 0.14, height: 0.48 },
        attributes: { helmet: false, mask: false, coverall: true },
      },
    ],
  },
}

export function listMockScenarios() {
  return Object.entries(SCENARIOS).map(([key, value]) => ({
    key,
    label: value.label,
  }))
}

export class MockDetectorAdapter extends DetectorAdapter {
  constructor(scenarioKey = 'one_person') {
    super()
    this.scenarioKey = scenarioKey
  }

  get id() {
    return 'mock-detector-v1'
  }

  get source() {
    return 'MOCK'
  }

  detect({
    cameraId,
    frameWidth = 1920,
    frameHeight = 1080,
    scenarioKey = this.scenarioKey,
    timestamp = new Date().toISOString(),
  }) {
    const scenario = SCENARIOS[scenarioKey]
    if (!scenario) {
      throw new Error(`Unknown mock scenario: ${scenarioKey}`)
    }

    return {
      cameraId,
      timestamp,
      source: this.source,
      frameWidth,
      frameHeight,
      objects: scenario.objects.map((item) => ({ ...item })),
    }
  }
}

export const defaultMockDetectorAdapter = new MockDetectorAdapter()

File: backend/app/core/detectors/mock_detector_adapter.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from app.core.detectors.capabilities import (
    DETECTOR_STATUS_FAILED,
    DETECTOR_STATUS_RUNNING,
    DETECTOR_STATUS_STARTING,
    DETECTOR_STATUS_STOPPED,
    DetectorCapabilities,
)
from app.core.detectors.detector_adapter import BaseDetectorAdapter
from app.core.event_bus import get_event_bus
from app.core.event_bus import event_types as topics
from app.core.observation_schema import DEFAULT_SCHEMA_VERSION
from app.services.observation_service import utc_now_iso

SCENARIOS: dict[str, dict[str, Any]] = {
    "one_person": {
        "objects": [
            {
                "trackId": "T-001",
                "class": "person",
                "confidence": 0.94,
                "bbox": {"x": 0.35, "y": 0.2, "width": 0.12, "height": 0.45},
                "attributes": {"helmet": True, "mask": True, "coverall": True},
            }
        ]
    },
    "three_persons": {
        "objects": [
            {
                "trackId": "T-101",
                "class": "person",
                "confidence": 0.91,
                "bbox": {"x": 0.1, "y": 0.25, "width": 0.1, "height": 0.4},
                "attributes": {"helmet": True, "mask": False, "coverall": True},
            },
            {
                "trackId": "T-102",
                "class": "person",
                "confidence": 0.88,
                "bbox": {"x": 0.35, "y": 0.22, "width": 0.11, "height": 0.42},
                "attributes": {"helmet": True, "mask": True, "coverall": True},
            },
            {
                "trackId": "T-103",
                "class": "person",
                "confidence": 0.86,
                "bbox": {"x": 0.62, "y": 0.28, "width": 0.1, "height": 0.38},
                "attributes": {"helmet": False, "mask": True, "coverall": True},
            },
        ]
    },
}


def _publish_detector_event(topic: str, detector_name: str, **extra: Any) -> None:
    get_event_bus().publish(
        topic,
        {
            "topic": topic,
            "timestamp": utc_now_iso(),
            "data": {"detectorName": detector_name, **extra},
        },
    )


class MockDetectorAdapter(BaseDetectorAdapter):
    def __init__(self) -> None:
        super().__init__(
            name="mock-detector-v1",
            source="MOCK",
            capabilities=DetectorCapabilities(
                person_detection=True,
                animal_detection=True,
                ppe_detection=True,
                tracking=True,
            ),
        )

    def start(self) -> None:
        self._health.status = DETECTOR_STATUS_STARTING
        self._health.message = "Starting mock detector"
        _publish_detector_event(topics.DETECTOR_STARTED, self.name)
        self._health.status = DETECTOR_STATUS_RUNNING
        self._health.started_at = utc_now_iso()
        self._health.message = "Mock detector running"

    def stop(self) -> None:
        self._health.status = DETECTOR_STATUS_STOPPED
        self._health.stopped_at = utc_now_iso()
        self._health.message = "Mock detector stopped"
        _publish_detector_event(topics.DETECTOR_STOPPED, self.name)

    def detect(self, context: dict[str, Any]) -> dict[str, Any]:
        scenario_key = context.get("scenarioKey") or context.get("scenario_key") or "one_person"
        scenario = SCENARIOS.get(scenario_key)
        if not scenario and context.get("objects"):
            scenario = {"objects": context["objects"]}
        if not scenario:
            raise ValueError(f"Unknown mock scenario: {scenario_key}")

        payload = {
            "cameraId": context["cameraId"],
            "timestamp": context.get("timestamp") or datetime.now(timezone.utc).isoformat(),
            "source": self.source,
            "schemaVersion": context.get("schemaVersion") or DEFAULT_SCHEMA_VERSION,
            "frameWidth": context.get("frameWidth") or 1920,
            "frameHeight": context.get("frameHeight") or 1080,
            "objects": [dict(item) for item in scenario["objects"]],
        }
        self._emit_observation(payload)
        return payload
